### PR TITLE
WIP: Maintain a raw encoding on decode for faster subsequent encode

### PR DESCRIPTION
**Pull request by brendandburns**
@lavalamp @caesarxuchao @davidopp @wojtek-t @gmarek 

This is a work in progress, but on my micro benchmark it reduces allocations in decode/encode by 30% and runtime by ~40-50%

If this looks reasonable, I will polish it up (in particular, make it optional, rather than always on, so that in-memory objects that we keep around (e.g in the kubelet) don't get bloated)

Also work out making sure the raw data doesn't get re-serialized.

### CODE REVIEW

File: pkg/api/unversioned/types.go [hunk 1]
```diff
@@ -33,6 +38,8 @@ type TypeMeta struct {
 	// may reject unrecognized values.
 	// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#resources
 	APIVersion string `json:"apiVersion,omitempty"`
+
+	Raw *RawData
```
Comment (lavalamp) [inline pkg/api/unversioned/types.go]: This must *not* be public, it can't show up in the API!

**Comment by lavalamp** [comment]
I think this can get the "hack of the year" award. :)

To make this work, you need to:
* modify ResourceVersion in the raw json (i.e., emit the first part, emit a `[, ]"resourceVersion": "X"` at the right place, then emit the rest of the raw json)
* make sure, when an object is modified, that Raw gets reset. Specifically I'm worried about copy-then-modify.

I think if you can do that, it should pass all the tests, but the second one looks hard.

**Comment by mvdan** [comment]
Instead of invalidating Raw by e.g. resetting it to nil, you could take a slightly different approach. You could check whether the data has been changed at encode time. For example, keeping a "checksum" of the data and checking against it.

The dumbest way to do such a checksum would be to make a hash out of the encoded json, which would mean you'd be back to running the json encode every time.

A slightly better version of that would be using `fmt.Sprintf("%v", data)`, which wouldn't contain the keys/names, so it should in theory be somewhat faster.

Probably the best way to do this would be to use `reflect` to navigate the data recursively and generate a checksum that doens't involve generating a large string with all the data, like `encoding/json` and `fmt` do.

I was thinking that we could use a trick to reuse what `map` does for structs as keys, but it only works with underlying types that are comparable (see "Key types" in https://blog.golang.org/go-maps-in-action).

**Comment by lavalamp** [comment]
We have a hashing function floating around somewhere that could be used,
but I'd be astonished if that didn't give back nearly all the speed
improvements.

**Comment by mvdan** [comment]
Just found it in `pkg/util/hash.go` - it seems to use a version of formatting `%#v` that dereferences pointers.

I do think that navigating a struct and taking note of its values in a hash should be noticeably faster than json marshaling. The hashing function in pkg/util could perhaps be improved, e.g. by hashing a binary representation of the data instead of a plaintext one.

**Comment by brendandburns** [comment]
Hrm, its worth testing, but we'd definitely have to generate the hash function, instead of using reflection...

**Comment by lavalamp** [comment]
> I do think that navigating a struct and taking note of its values in a hash should be noticeably faster than json marshaling.

I doubt it. Reflection is the vast majority of json generation time. The difference between that and hashing is whether the output data goes to different bytes, or to the same bytes. You probably don't even get to skip outputting the keys.

That said, I'll take it back if you produce a microbenchmark that shows me wrong. I'm guessing the hashing is not more than 10% faster, and I'll be surprised if it's more than 5%...

> generate the hash function

This optimization is not worth that much code complexity, IMO

File: pkg/api/unversioned/types.go (hunk 1 above)
Comment (brendandburns) [inline pkg/api/unversioned/types.go]: fwiw, go doesn't support reflection on non-exported fields, even within the same package, so this needs to be exported.

Comment (lavalamp) [inline pkg/api/unversioned/types.go]: I think it's kind of a non-starter to let an optimization like this leak nonsense into our API. I'm less willing to take this than even super gross things like using the unsafe package to turn a runtime.Object into a uintptr, which we use as the key to a map with encodings. (!)

**Comment by mvdan** [comment]
@lavalamp this was still bugging me, so I played with it for a bit. The findings are interesting.

```
 % go test -bench=.
PASS
BenchmarkGob-4             50000             33093 ns/op
BenchmarkJson-4            10000            233827 ns/op
BenchmarkKubeNoTypes-4      5000            332162 ns/op
BenchmarkKube-4             5000            395548 ns/op
ok      github.com/mvdan/hashtest       8.068s
```

Each of these means feeding different representations of structs to the hasher. Gob and Json are plain `encoding/gob` and `encoding/json`, Kube is what you do in `pkg/util/hash.go` with go-spew and KubeNoTypes is a stripped down version of Kube (removing types and keys).

We can take Json as the starting point, since that's the cost we want to avoid in this PR. Even when stripped down, your hash function - namely the go-spew usage - is too expensive. I would need to investigate further, but my thinking is that go-spew itself is not very efficient.

Since `encoding/json` was so noticeably faster, I wanted to play with other encoding options in the standard library. Since `encoding/gob` isn't human readable, I thought it could perform better. And it does - to the point where it's ten times faster than KubeNoTypes.

If you look at the tests I ran, I'm checking that all of those do dereference pointers when using reflection. And the benchmark test is a large struct with pointers, since my intention was to see how each method would deal with large structures and reflection.

Source is at https://github.com/mvdan/hashtest.

I'm thinking that we should clearly move `pkg/util/hash.go` to `encoding/gob`, unless having the specific types is important. For json caching it definitely isn't, since `encoding/json` omits that info too. All that json does is keys and values, which gob does just as fine.

I should confess I'm cheating though - If you look at https://github.com/golang/go/issues/7872, you'll see that `encoding/json` buffers the whole thing and writes only once. `encoding/gob` does writes in chunks, so that really shows since my benchmark struct is large. This should all be better once `encoding/json` is fixed, but that won't happen until at least 1.6.

**Comment by mvdan** [comment]
I forgot to make a point:

> Reflection is the vast majority of json generation time.

That should be true, but isn't true in Go 1.5 as per the benchmarks above. If `gob` can do it 7x faster than `json`, that means reflection is at most 15% of `json`'s runtime.

**Comment by mvdan** [comment]
Disregard the `encoding/gob` numbers above, I was ignoring errors (yes, I had it coming). I'll be back with better numbers.

**Comment by mvdan** [comment]
Here they are:

```
 % go test -bench=. -benchtime=2s
PASS
BenchmarkGob-4             10000            254616 ns/op
BenchmarkJson-4            10000            230740 ns/op
BenchmarkJsonUgorji-4      10000            216633 ns/op
BenchmarkKubeNoTypes-4     10000            335086 ns/op
BenchmarkKube-4            10000            399854 ns/op
ok      github.com/mvdan/hashtest       14.534s
```

I was hoping that ugorji would be fast enough for caching to not make sense. Looks like generating code to work around reflection did not improve things much.

**Comment by wojtek-t** [comment]
@mvdan - thanks for those experiments

Regarding ugorji - I'm already working on making it work - see https://github.com/kubernetes/kubernetes/pull/15072
I'm blocked on some small things there, but I'm mostly done - so don't duplicate my work around it.

**Comment by wojtek-t** [comment]
>
I was hoping that ugorji would be fast enough for caching to not make sense. Looks like generating code to work around reflection did not improve things much.

I don't agree that it didn't improve things much - it improved it by 2x which I think is much. But yeah - we still need caching.

**Comment by wojtek-t** [comment]
It seems I misread your results. Anyway - where do they come from? When I was running some experiments I observed around 2x improvement ugorji vs encoding/json ?

**Comment by mvdan** [comment]
@wojtek-t that's precisely why I added ugorji, to see what difference it would make.

What I get here is a 10% improvement. If you look at the struct I'm using, I used a slice with 100 elements with interfaces, to try and see how they would perform when the data is large and there is reflection involved. I would think it would be in this scenario where ugorji would perform even better.

Note: I did not use unsafe with ugorji.

**Comment by wojtek-t** [comment]
>
What I get here is a 10% improvement. If you look at the struct I'm using, I used a slice with 100 elements with interfaces, to try and see how they would perform when the data is large and there is reflection involved. I would think it would be in this scenario where ugorji would perform even better.

I don't think we should come up with dedicated structs to check. 
What I've done is I took few random pods, nodes and rc from our test clusters and tried encoding on those objects. And difference on that objects were around 2x.

>
Note: I did not use unsafe with ugorji.

Me neither.

**Comment by mvdan** [comment]
Do you have these random pod/node/rc objects somewhere?

**Comment by wojtek-t** [comment]
For some of them you can look into:
https://github.com/kubernetes/kubernetes/tree/master/pkg/api
for *_example.json files.

**Comment by lavalamp** [comment]
@mvdan thanks for following up with actual data. Looks like you got something 2x faster than our current hash function, and it's using the thing that we intend to use to encode/decode structs anyway, so using it to hash them to see if we can reuse a previous run doesn't make too much sense :)

I do share @wojtek-t's concern that maybe different test data would show different results...

Might be worth changing our hash function anyway, but that's a different issue :)

@brendandburns I'm going to close this, it's a lot of complexity. Let's revisit if we still think we need this now that @wojtek-t has got ugorji in.

**Pull request closed without merging**